pkg/container/container.go

```diff
@@ -72,13 +96,121 @@ type container struct {
 	id         string
 	name       string
 	privileged bool
+	spec       generate.Generator
 }
 
 // New creates a new, empty Sandbox instance
-func New(ctx context.Context) Container {
+func New(ctx context.Context) (Container, error) {
+	spec, err := generate.New("linux")
+	if err != nil {
+		return nil, err
+	}
 	return &container{
-		ctx: ctx,
+		ctx:  ctx,
+		spec: spec,
+	}, nil
+}
+
+// SpecAddMount adds a specified mount to the spec
+func (c *container) SpecAddMount(destination, mntType, sourceType, mountToRemove string, options []string) {
+	c.spec.RemoveMount(mountToRemove)
+	mnt := rspec.Mount{
+		Destination: destination,
+		Type:        mntType,
+		Source:      sourceType,
+		Options:     options,
+	}
+	c.spec.AddMount(mnt)
+}
+
+// SpecAddAnnotation adds all annotations to the spec
+func (c *container) SpecAddAnnotations(sb *sandbox.Sandbox, containerVolumes []oci.ContainerVolume, mountPoint, configStopSignal string, imageResult *storage.ImageResult, isSystemd, systemdHasCollectMode bool) (err error) {
+	const podTerminationGracePeriodLabel = "io.kubernetes.pod.terminationGracePeriod"
+	kubeAnnotations := c.Config().GetAnnotations()
+	created := time.Now()
+	labels := c.Config().GetLabels()
+
+	image, err := c.Image()
+	if err != nil {
+		return err
+	}
+	logPath, err := c.LogPath(sb.LogDir())
+	if err != nil {
+		return err
+	}
+	c.spec.AddAnnotation(annotations.Image, image)
+	c.spec.AddAnnotation(annotations.ImageName, imageResult.Name)
+	c.spec.AddAnnotation(annotations.ImageRef, imageResult.ID)
+	c.spec.AddAnnotation(annotations.Name, c.Name())
+	c.spec.AddAnnotation(annotations.ContainerID, c.ID())
+	c.spec.AddAnnotation(annotations.SandboxID, sb.ID())
+	c.spec.AddAnnotation(annotations.SandboxName, sb.Name())
+	c.spec.AddAnnotation(annotations.ContainerType, annotations.ContainerTypeContainer)
+	c.spec.AddAnnotation(annotations.LogPath, logPath)
+	c.spec.AddAnnotation(annotations.TTY, fmt.Sprintf("%v", c.Config().Tty))
+	c.spec.AddAnnotation(annotations.Stdin, fmt.Sprintf("%v", c.Config().Stdin))
+	c.spec.AddAnnotation(annotations.StdinOnce, fmt.Sprintf("%v", c.Config().StdinOnce))
```

Comment:
I know this is not something new but rather copied from the old code, and yet while at it can you switch to using `strconv.FormatBool` for these ^^^ three, as fmt.Sprintf is pretty heavy.

Reply:
fixed!